lib/TPP/ConvertPerfToFunc.cpp

```diff
@@ -0,0 +1,361 @@
+  auto deltas = rewriter.create<memref::CastOp>(loc, memRefType, buff);
+
+  // Iterate over the whole buffer and compute the standard deviation of the
+  // measured time delta values.
+  // Implemented directly as scf to keep further lowering simple.
+  auto zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+  auto one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+  auto len = rewriter.create<memref::DimOp>(loc, deltas, zero);
+  auto floatType = rewriter.getF64Type();
+  auto result = rewriter.create<arith::ConstantFloatOp>(
+      loc, APFloat::getZero(floatType.getFloatSemantics()), floatType);
+  auto mean = block->getArguments()[1];
+
+  auto loopNest = scf::buildLoopNest(
+      rewriter, loc, /*lbs=*/ValueRange{zero}, /*ubs=*/ValueRange{len},
+      /*steps=*/ValueRange{one}, ValueRange{result},
+      [&](OpBuilder &b, Location loc, ValueRange localIvs,
+          ValueRange iterArgs) -> scf::ValueVector {
+        auto meas = rewriter.create<memref::LoadOp>(loc, deltas, localIvs);
+        auto delta = rewriter.create<arith::SubFOp>(loc, meas, mean);
+        auto deltaSqr = rewriter.create<arith::MulFOp>(loc, delta, delta);
+        auto sum = rewriter.create<arith::AddFOp>(loc, deltaSqr, iterArgs[0]);
+
+        return scf::ValueVector({sum});
+      });
+  assert((loopNest.results.size() == 1) && "expect only 1 loop result");
+
+  // Compute standard deviation.
+  auto lenInt = rewriter.create<arith::IndexCastOp>(
+      loc, rewriter.getIntegerType(64), len);
+  auto size = rewriter.create<arith::SIToFPOp>(loc, floatType, lenInt);
+  auto variance =
+      rewriter.create<arith::DivFOp>(loc, floatType, loopNest.results[0], size);
+  auto stdev = rewriter.create<math::SqrtOp>(loc, variance);
+
+  // Return the computed stdev value.
+  rewriter.create<func::ReturnOp>(loc, ValueRange{stdev});
+}
+
+// Creates function prototypes and insert calls to the perf runtime functions.
+static LogicalResult buildPerfFuncCall(Location loc, std::string funcName,
+                                       Operation *op,
+                                       PatternRewriter &rewriter) {
+  if (op->getNumResults() > 1)
+    return op->emitError(
+               "expected operation to have 0 or 1 result, but provided ")
+           << op->getNumResults();
+
+  FlatSymbolRefAttr fnName = SymbolRefAttr::get(op->getContext(), funcName);
+  ModuleOp module = op->getParentOfType<ModuleOp>();
+
+  // Create function prototype if it is not available yet.
+  if (!module.lookupSymbol(fnName.getAttr())) {
+    OpBuilder::InsertionGuard guard(rewriter);
+    // Insert before module terminator.
+    rewriter.setInsertionPoint(module.getBody(),
+                               std::prev(module.getBody()->end()));
+
+    auto libFnType = rewriter.getFunctionType(
+        extractNormalizedTypes(rewriter, op->getOperands()),
+        extractNormalizedTypes(rewriter, op->getResults()));
+    func::FuncOp funcOp =
```

Comment:
This disconnects the function from the builder and creates coupling between the creation outside of `buildPerfMeanFunc` and adding the entry block inside.

This kind of code generation is best if the functions are declared/defined outline and the function symbol is kept on a map to be used by code generation.

`Sink` is the only one that needs to be "on demand" because of the argument type, but that's an empty function, so it's the trivial case.

Reply:
The builders should be more self contained now.

I don't think cross-module `func.call`s are really supported so I gave up on the LUT approach as you could have perf ops present in different modules.